### ISSUE TITLE
Support for "grouped" or indexed draws_df

### ISSUE THREAD

**Issue by jburos** [stan-dev/posterior#112]
This may not be common usage, but I will ask in case others have this issue or it's already solved but I'm not aware of it.

We are often working with data in a `draws_df`-like structure, after reading in from netcdf (per my comment in https://github.com/stan-dev/posterior/issues/85#issuecomment-786932435). This data structure looks identical to the result of having using `tidybayes::spread_draws` to organize results into a long/tidy structure.

Here is an example to generate a similar data structure for a 1d parameter:

```r
library(tibble)
library(rstan)

data <- tibble::tibble(school = letters[1:8],
                       y = c(28.39, 7.94, -2.75 , 6.82, -0.64, 0.63, 18.01, 12.16),
                       sigma = c(14.9, 10.2, 16.3, 11.0, 9.4, 11.4, 10.4, 17.6))

stan_data <- list(J = nrow(data), y = data$y, sigma = data$sigma)
model_code <- 'data {
  int<lower = 0> J; // number of schools 
  vector[J] y; // estimated treatment effects
  vector<lower = 0>[J] sigma; // standard errors
}

parameters {
  real mu; // pop mean
  real<lower = 0> tau; // pop std deviation
  vector[J] eta; // school-level errors
}

transformed parameters {
  vector[J] theta = mu + tau * eta; // school effects
}

model {
  eta ~ normal(0, 1);
  y ~ normal(theta, sigma);
}'

fit <- rstan::stan(model_code = model_code, data = stan_data, chains = 1, iter = 200)
```

And I extracted by draws like so:

```r
theta_draws <- tidybayes::spread_draws(fit, theta[school_id])
```

The resulting dataset looks a like this:

```r
> theta_draws
# A tibble: 800 x 5
# Groups:   school_id [8]
   school_id theta .chain .iteration .draw
       <int> <dbl>  <int>      <int> <int>
 1         1 12.0       1          1     1
 2         1  3.48      1          2     2
 3         1  7.20      1          3     3
 4         1  6.55      1          4     4
 5         1  6.55      1          5     5
 6         1  7.76      1          6     6
 7         1  5.12      1          7     7
 8         1  8.62      1          8     8
 9         1  7.88      1          9     9
10         1  6.64      1         10    10
```

I can convert this to a `draws_df`, like so:

```r
> theta_draws %>% dplyr::select(-.draw) %>% posterior::as_draws_df()
# A draws_df: 800 iterations, 1 chains, and 2 variables
   school_id theta
1          1  12.0
2          2   4.2
3          3   7.9
4          4   7.9
5          5   6.4
6          6   4.4
7          7  10.5
8          8   3.2
9          1   3.5
10         2   6.1
# ... with 790 more draws
# ... hidden reserved variables {'.chain', '.iteration', '.draw'}
```

However, this results in the `school_id` being considered a variable, which is not quite right. And this is not the right structure for summarizing quantities like `rhat` / etc. 

Right now I'm doing something like the following:

```r
> theta_draws %>% select(-.draw) %>% group_by(school_id) %>% tidyr::nest() %>% dplyr::mutate(draws = purrr::map(data, posterior::as_draws_df)) %>% dplyr::mutate(rhat = purrr::map_dbl(draws, posterior::rhat_basic))
# A tibble: 8 x 4
# Groups:   school_id [8]
  school_id data               draws                 rhat
      <int> <list>             <list>               <dbl>
1         1 <tibble [100 × 3]> <draws_df [100 × 4]>  3.04
2         2 <tibble [100 × 3]> <draws_df [100 × 4]>  3.06
3         3 <tibble [100 × 3]> <draws_df [100 × 4]>  3.06
4         4 <tibble [100 × 3]> <draws_df [100 × 4]>  3.11
5         5 <tibble [100 × 3]> <draws_df [100 × 4]>  3.15
6         6 <tibble [100 × 3]> <draws_df [100 × 4]>  3.11
7         7 <tibble [100 × 3]> <draws_df [100 × 4]>  3.02
8         8 <tibble [100 × 3]> <draws_df [100 × 4]>  3.11
```

Is there a better solution to this? Curious if you are thinking about supporting a grouped or denormalized data structure for storing Variables.

**Comment by paul-buerkner** [comment]
I think the way currently intended by posterior is to have a separate column (variable) per school, as (I assume) theta is a different parameter per school.

**Comment by mjskay** [comment]
I think this workflow could be simplified slightly using the (relatively recent) `dplyr::group_modify()` function.

Using this data as an example:

```r
df = as_draws_df(data.frame(`x[1]` = rnorm(100), `x[2]` = rnorm(100), check.names = FALSE))
df
```
```
# A draws_df: 100 iterations, 1 chains, and 2 variables
     x[1]   x[2]
1  -0.805 -2.001
2   0.010 -0.489
3   0.120  0.807
4  -1.316  0.418
5  -1.537 -0.905
6   0.426  1.692
7  -1.602  0.614
8  -0.068 -0.348
9  -0.201 -0.990
10  0.468 -0.042
# ... with 90 more draws
# ... hidden reserved variables {'.chain', '.iteration', '.draw'}
```

`group_modify()` works similarly to the old `dplyr::do()`, letting you return an entire data frame and binding the results together. By default it leaves grouping variables out of the data frame it passes down to the function you give it, which means you can use the data frame passed down with `summarise_draws()`. The only annoying bit is the need to drop the `.draw` column, which I think should probably be changed (I filed an issue here: #114):

```r
df %>%
  spread_draws(x[i]) %>%
  select(-.draw) %>%
  group_modify(~ summarise_draws(.x))
```
```
# A tibble: 2 x 11
# Groups:   i [2]
      i variable     mean median    sd   mad    q5   q95  rhat ess_bulk ess_tail
  <int> <chr>       <dbl>  <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>    <dbl>    <dbl>
1     1 x        -0.00743 0.0227 0.954 0.895 -1.56  1.28 0.993    113.      102.
2     2 x         0.0815  0.0806 1.06  0.830 -1.66  2.04 1.03      92.1     108.
```

And if you just want specific summaries:

```r
df %>%
  spread_draws(x[i]) %>%
  select(-.draw) %>%
  group_modify(~ summarise_draws(.x, rhat_basic, ess_bulk))
```
```
# A tibble: 2 x 4
# Groups:   i [2]
      i variable rhat_basic ess_bulk
  <int> <chr>         <dbl>    <dbl>
1     1 x             0.993    113. 
2     2 x             1.02      92.1
```

If #114 is fixed you'll be able to skip the `select(-.draw)` call. The remaining question (to me) is if it makes sense to have an implementation of `summarise_draws` for the `grouped_df` class, which could basically do the above group_modify and allow you to do something like:

```r
df %>%
  spread_draws(x[i]) %>%
  summarise_draws(rhat_basic, ess_bulk)
```

I think it could be nice. I'd be curious what @paul-buerkner thinks.

**Comment by paul-buerkner** [comment]
I like your proposed workflow. i think supporting the grouped_df class in `summarize_draws` sounds like a good idea in principle. Where does this class come from exactly and how much code-overhead do you think is produced by supporting it?

**Comment by mjskay** [comment]
`grouped_df` is a class from dplyr that is a subclass of tibble that's added whenever someone uses `dplyr::group_by()` on a data.frame (see https://dplyr.tidyverse.org/reference/grouped_df.html).

I think the overhead won't be much --- an implementation of `summarise_draws.grouped_df()` should be just a couple of lines, essentially doing what I did above with `group_modify()`, maybe explicitly using `as_draws_df()` on each sub-table and calling down to `summarise_draws.draws_df()`. If you think it's a good idea I can draft a solution and combine it with a PR for #114?

**Comment by mjskay** [comment]
Ah hrm, there is one wrinkle: it would add a dplyr dependency, which might be undesirable. Another option (if that is not desirable) would be to put the implementation of `summarise_draws.grouped_df()` in tidybayes, since it already has a dplyr dependency and (I assume?) most people who would want it would already be using tidybayes.

**Comment by paul-buerkner** [comment]
I would tend to agree that this method might be better located in tidybayes than in posterior given the constraints you mentioned.

**Comment by mjskay** [comment]
okay cool, once #116 is merged I'll add an implementation to tidybayes (I have one ready to go) and then close this.

**Comment by mjskay** [comment]
@jburos the need to do `select(-.draw)` for this should be fixed now.

I've also added an implementation of `summarise_draws()` for grouped data frames to the *posterior* branch of tidybayes (see [here](https://github.com/mjskay/tidybayes/blob/posterior/R/summarise_draws.R)). That branch will eventually become the mainline tidybayes version after posterior reaches CRAN. In the meantime you can get it via:

```r
devtools::install_github("mjskay/tidybayes@posterior")
```

Note that that branch of tidybayes currently also depends on the *rv-like* branch of posterior, though that should (I hope?) soon be merged into master.
